uefi/src/mem/aligned_buffer.rs

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use alloc::alloc::{alloc, dealloc, Layout, LayoutError};
+use core::error::Error;
+use core::fmt;
+
+/// Helper class to maintain the lifetime of a memory region allocated with a non-standard alignment.
+/// Facilitates RAII to properly deallocate when lifetime of the object ends.
+///
+/// Note: This uses the global Rust allocator under the hood.
+#[allow(clippy::len_without_is_empty)]
+#[derive(Debug)]
+pub struct AlignedBuffer {
+    ptr: *mut u8,
+    layout: Layout,
+}
+
+impl AlignedBuffer {
+    /// Allocate a new memory region with the requested len and alignment.
+    pub fn alloc(len: usize, alignment: usize) -> Result<Self, LayoutError> {
+        let layout = Layout::from_size_align(len, alignment)?;
+        let ptr = unsafe { alloc(layout) };
+        Ok(Self { ptr, layout })
+    }
+
+    /// Get a pointer to the aligned memory region managed by this instance.
+    #[must_use]
+    pub const fn ptr(&self) -> *const u8 {
+        self.ptr.cast_const()
+    }
+
+    /// Get a mutable pointer to the aligned memory region managed by this instance.
+    #[must_use]
+    pub fn ptr_mut(&mut self) -> *mut u8 {
+        self.ptr
+    }
+
+    /// Get the size of the aligned memory region managed by this instance.
+    #[must_use]
+    pub const fn len(&self) -> usize {
+        self.layout.size()
+    }
+
+    /// Fill the aligned memory region with data from the given buffer.
+    pub fn copy_from(&mut self, data: &[u8]) {
+        let len = data.len().min(self.len());
```

Comment:
The std API usually doesn't do this kind of automatic length adjustment for slices of different lengths (e.g. https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice panics for different lengths). When you use this in the scsi code, I think the allocated buffer is the same length as the data to be copied right? So let's just `assert_eq!(self.len(), data.len())`.